### PR TITLE
feat: raise podcast page episode limit + add title search (#291)

### LINKED ISSUE #291: feat: raise podcast page episode limit and add client-side title search
## Context

**User Story:** As a podcast listener browsing a show's detail page, I want to see more than the last 20 episodes and filter them by title, so that I can find specific older episodes without scrolling blindly or giving up.

Today, `/podcast/[id]` caps the episode list at 20 (PodcastIndex-sourced) or 50 (RSS-sourced from our DB). Users who remember a specific episode from months or years ago have no way to reach it. PodcastIndex's `/episodes/byfeedid` endpoint has no search and no offset pagination, so a deep-archive solution requires either raising the fetch window or investing in a DB-backed sync (deferred — see follow-up note at bottom).

This issue covers the cheap, fast v1: raise the window to 200 episodes and add a client-side title filter. The full design and trade-off analysis are captured in the approved spec at [`.dev/pm/specs/2026-04-17-podcast-page-more-episodes.md`](.dev/pm/specs/2026-04-17-podcast-page-more-episodes.md) — agents picking up this issue should read it first.

## Complexity Hint

**Subsystems touched:** podcast detail page (server component) + episode-list component (client component)
**Independent stories:** 1
**Structural changes needed:** No — raises a constant and adds local client state; no interface changes
**Scope note:** none detected

## Proposed Solution

1. Raise the per-podcast episode fetch limit from 20 → 200 on both code branches of the podcast detail page (PodcastIndex-sourced and RSS-sourced-from-DB). Extract the limit into a named constant at the top of the file so the two branches stay in sync.
2. Add a search input above the episode list inside the existing `EpisodeList` client component. Filter the rendered episode array by case-insensitive substring match on `episode.title`. Empty query renders all episodes unchanged.
3. When the query is non-empty and produces zero results but the source array is non-empty, render a `"No episodes match '<query>'"` empty state instead of the existing `"No episodes found for this podcast."` one.
4. Hide the search input entirely when `episodes.length === 0` (nothing to search).

## Acceptance Criteria

- [ ] VERIFY: Visiting `/podcast/[id]` for a PodcastIndex-sourced podcast fetches up to 200 episodes (previously 20)
- [ ] VERIFY: Visiting `/podcast/[id]` for an RSS-sourced podcast (id prefixed with `rss-`) reads up to 200 episodes from the DB (previously 50)
- [ ] VERIFY: Episode count shown in `Episodes (N)` header reflects the loaded total, not the filtered subset
- [ ] VERIFY: A search input renders above the episode list when `episodes.length > 0` and is hidden when the list is empty
- [ ] VERIFY: Typing in the search input filters episodes by case-insensitive substring match on `title`, with no network request
- [ ] VERIFY: Clearing the search input restores the full list
- [ ] VERIFY: A non-empty query with zero matches renders a `"No episodes match '<query>'"` message
- [ ] VERIFY: Existing "No episodes found for this podcast." state still renders when the source array is empty
- [ ] VERIFY: Unit tests cover: match found, case-insensitive match, no-match empty state, clear-filter restores list, search hidden when no episodes
- [ ] VERIFY: `bun run lint`, `bun run test`, and `bun run build` all pass

## Edge Cases

- **Podcast has >200 episodes:** Only the 200 most recent are loaded — accepted cap for v1 until Approach C ships. The filter only searches within that window.
- **Podcast has 0 episodes:** Search input is hidden; existing empty state renders.
- **Query with whitespace only (e.g. `"   "`):** Treat as empty query (trim before comparing); render full list.
- **Query with special regex characters (e.g. `"C++"`, `"(`, `".*"`)**: Use plain substring match (`String.prototype.includes`), not regex — no special-character handling needed and no injection risk.
- **Very long episode titles:** Match logic is substring; no truncation concern at match time. Rendering truncation is already handled by `EpisodeCard`.

### Files to Modify

- `src/app/(app)/podcast/[id]/page.tsx` — extract a `PODCAST_PAGE_EPISODE_LIMIT = 200` constant near the top. Update line 69 (`limit: 50` in `loadRssPodcast`) and line 268 (`getEpisodesByFeedId(feedId, 20)`) to both use the constant.
- `src/components/podcasts/episode-list.tsx` — add local `useState<string>("")` for the query, add a shadcn `<Input>` with a search icon and placeholder `"Search episodes by title…"`, filter the `episodes` array with `.filter(e => e.title.toLowerCase().includes(query.trim().toLowerCase()))` before mapping, and handle the no-match empty state.
- `src/components/podcasts/__tests__/episode-list.test.tsx` — new test file (if absent) or extend existing. Cover: renders all when query empty; filters case-insensitively; shows no-match state on empty filtered result; restores list on clear; hides input when `episodes` prop is empty.

### Files to Create

None — all changes land in existing files.

### Approach

1. Open `src/app/(app)/podcast/[id]/page.tsx`. Add `const PODCAST_PAGE_EPISODE_LIMIT = 200;` near the top of the file (after imports, before `isRssSourced`). Replace `limit: 50` at line 69 with `limit: PODCAST_PAGE_EPISODE_LIMIT`. Replace `20` at line 268 with `PODCAST_PAGE_EPISODE_LIMIT`.
2. Open `src/components/podcasts/episode-list.tsx`. Import `useState` from `react` and `Input` from `@/components/ui/input` (run `bunx shadcn@latest add input` if the component isn't already present — check `src/components/ui/` first). Optionally import a search icon from `lucide-react` (`Search`).
3. Inside the component, add `const [query, setQuery] = useState("")`. Compute `const normalizedQuery = query.trim().toLowerCase()` and `const filteredEpisodes = normalizedQuery ? episodes.filter(e => e.title.toLowerCase().includes(normalizedQuery)) : episodes`.
4. Render the search input above the episode map, only when `!isLoading && !error && episodes.length > 0`. Render the existing "No episodes found" state when the source array is empty; render a new "No episodes match '<query>'" state when `normalizedQuery` is non-empty and `filteredEpisodes.length === 0`.
5. Map over `filteredEpisodes` (not `episodes`) in the render output.
6. Add/extend tests under `src/components/podcasts/__tests__/episode-list.test.tsx`. Use React Testing Library (`render`, `screen`, `fireEvent` or `userEvent`) following the project's existing test patterns.
7. Run `bun run lint`, `bun run test`, and `bun run build` locally. Fix any issues. The pre-commit hook will re-run lint and tests on commit.
8. Update `CHANGELOG.md` under `[Unreleased]` → `Changed` (e.g., "Podcast detail page now shows up to 200 episodes and supports client-side title search.").

### Patterns to Follow

- `EpisodeList` is already a `"use client"` component (see the directive at the top of `src/components/podcasts/episode-list.tsx`) — no need to change boundaries.
- Use shadcn `<Input>` for the search field (see `AGENTS.md` — "shadcn/ui components live in `src/components/ui/`").
- Use the `@/` path alias for imports, not relative paths.
- Follow TypeScript strict mode and the existing functional-component style in the file.

### Constraints

- **Must not trigger a network request on every keystroke** — filtering is purely client-side over the already-loaded `episodes` prop.
- **Must not change the `EpisodeList` public prop shape** — callers continue to pass `episodes`, `statusMap`, `scoreMap` unchanged. Adding a local search state is an internal concern.
- **Must not regress SSR / server-component behavior on the page** — the page itself stays server-rendered; only `EpisodeList` (already client) owns the new state.

### In Scope

- Raise episode fetch limit to 200 on both branches of the podcast detail page (extract to a named constant).
- Add client-side title search input + filter logic inside `EpisodeList`.
- "No matches" empty state for non-empty queries with zero results.
- Unit tests covering filter match/no-match/clear/case-insensitive/hidden-when-empty.
- `CHANGELOG.md` entry under `[Unreleased] → Changed`.

### Out of Scope

- **DB-backed episode backfill + server-side search** — tracked as a follow-up (Approach C in the spec). Removes the 200-episode cap and enables search for arbitrarily deep archives.
- Server-side pagination or "load more" buttons.
- Filtering on description, author, publish date, or summary status.
- URL-synced search state (no `nuqs` integration).
- Full-text indexing, fuzzy matching, or typo tolerance.
- Changes to the dashboard, library, or search pages (they manage their own episode windows and are not part of this story).

## Dependencies

- Blocked by: none
- Blocks: none
- Requires: shadcn `Input` component — if `src/components/ui/input.tsx` does not exist, run `bunx shadcn@latest add input` as part of implementation.

## Testing Requirements

- [ ] Unit tests in `src/components/podcasts/__tests__/episode-list.test.tsx` for the filter behavior
- [ ] Manual smoke test on a PodcastIndex-sourced podcast with >50 episodes to confirm the window raised
- [ ] Manual smoke test on an RSS-sourced podcast to confirm the DB query limit raised
- [ ] Edge case tests for: empty query, whitespace-only query, no-match query, special characters in query, empty episodes array

---

## Spec Reference

Full design rationale, alternatives considered (Approach B "load older" pagination, Approach C DB backfill), and trade-off analysis are in the approved spec:

**Spec:** `.dev/pm/specs/2026-04-17-podcast-page-more-episodes.md`

**Follow-up (not part of this issue):** Approach C — DB-backed episode archive + server-side search. Sync all episodes for a podcast into the `episodes` table on first visit (background job), then read from the DB with proper pagination and Postgres-backed title search. Worth revisiting after observing real usage of this v1 — if users regularly hit the 200-episode cap, Approach C is justified.

### PR DESCRIPTION
**Pull request by funny-otter**
Closes #291.

## Summary
- Raise per-podcast episode fetch cap from 20 (PodcastIndex) / 50 (RSS) to 200 via a single shared `PODCAST_PAGE_EPISODE_LIMIT` constant in `src/app/(app)/podcast/[id]/page.tsx`.
- Add a case-insensitive client-side title filter inside the existing client component `EpisodeList`, with a leading search icon, accessible label, and a `No episodes match "<query>"` empty state for non-matching queries.
- Defensive null-coercion `(episode.title ?? "").toLowerCase()` in the filter — the type claims `string`, but PodcastIndex is an external boundary; the larger 200-record window materially raises the chance of hitting a malformed payload, and a client-side `TypeError` would not be caught by the page's server-side try/catch.
- Hide the search input entirely when the episode list is empty, while loading, or while erroring (early returns preserve existing behavior).
- Whitespace-only queries are treated as empty (no spurious "no match" state).
- New Storybook story (`episode-list.stories.tsx`) covering Default / Loading / Error / Empty / SingleEpisode.
- New unit tests (`__tests__/episode-list.test.tsx`) covering 11 cases: empty query, case-insensitive match, mixed-case, no-match empty state, clear-restores, whitespace-only, source-empty hides input, isLoading hides input, error hides input, statusMap/scoreMap pass-through preserved post-filter, and a regression test for null titles.
- CHANGELOG entry under `[Unreleased] → Changed`.

Out of scope (per issue): server-side pagination, DB-backed episode archive (Approach C in the spec), URL-synced query state, fuzzy matching.

## Test plan
- [x] `bun run lint` — clean
- [x] `bun run test` — 1739 / 1739 pass (138 files, +4 new)
- [x] `bun run build` — succeeds
- [x] Storybook + agent-browser end-to-end: search renders above 5 episodes, `react` narrows 5 → 1, `RuSt` matches case-insensitively, `xyz-no-match` shows the no-match empty state, backspace-clear restores the full list, Empty story hides the search input.
- [ ] Manual smoke on a PodcastIndex-sourced podcast with > 50 episodes (loaded from PodcastIndex) — confirms the 20 → 200 raise.
- [ ] Manual smoke on an RSS-sourced podcast (`rss-…` id) with > 50 DB episodes — confirms the 50 → 200 raise.

## Summary by CodeRabbit

* **New Features**
  * Podcast detail pages now load up to 200 episodes and include a client-side title search to filter the loaded episodes.
* **Tests**
  * Added unit tests covering episode list rendering, search/filter behavior, empty states, loading/error states, and status/score wiring.
* **Chores**
  * Added Storybook stories demonstrating EpisodeList states.
* **Documentation**
  * Updated changelog to describe the expanded episode loading and new search.

### CODE REVIEW

File: src/components/podcasts/episode-list.tsx [hunk 1]
```diff
@@ -1,6 +1,9 @@
 "use client";
 
+import { useState } from "react";
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/components/podcasts/episode-list.tsx]: Fixed: added `useMemo` import alongside `useState`.

File: src/components/podcasts/episode-list.tsx [hunk 2]
```diff
@@ -42,11 +47,47 @@ export function EpisodeList({ episodes, isLoading, error, statusMap, scoreMap }:
     );
   }
 
+  const trimmedQuery = query.trim();
+  const normalizedQuery = trimmedQuery.toLowerCase();
+  const filteredEpisodes = normalizedQuery
+    ? episodes.filter((episode) =>
+        (episode.title ?? "").toLowerCase().includes(normalizedQuery),
+      )
+    : episodes;
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/components/podcasts/episode-list.tsx]: Fixed: wrapped the filter in `useMemo` keyed on `episodes` and `normalizedQuery`. Moved the hook above the early returns so it isn't called conditionally.

**Comment by funny-otter** [comment]
> ## Code Review
> 
> This pull request increases the episode limit to 200 and adds a client-side search bar to filter episodes by title, supported...

Answered: the `useMemo` suggestion is implemented — filter now runs through `useMemo([episodes, normalizedQuery])`, with the hook placed before the early returns to satisfy the Rules of Hooks.

**Comment by funny-otter** [comment]
> 🧹 Nitpick comments (3)
> `src/components/podcasts/episode-list.stories.tsx` — Minor: `Error` as a story export shadows the global `Error` constructor...

Acknowledged — deferring. The shadowing is harmless in this file (no `throw new Error(...)` anywhere), and renaming the story exports would invalidate the freshly-regenerated VRT baselines and require another baseline-regen cycle. Happy to rename in a follow-up PR alongside a sweep of other story files if we want to standardize on `*State` naming.

**Comment by funny-otter** [comment]
## PR Comment Status

| Status | Threads | Review Bodies | Issue Comments | Total |
|--------|---------|---------------|----------------|-------|
| Resolved | 0 | 0 | 2 | 2 |
| Fixed by DLC | 2 | 0 | 0 | 2 |
| Answered by DLC | 0 | 1 | 0 | 1 |
| Skipped (user decision) | 0 | 0 | 0 | 0 |
| Discussion-Deferred | 0 | 1 | 0 | 1 |
| Discussion-Tracked | 0 | 0 | 0 | 0 |
| Blocked | 0 | 0 | 0 | 0 |
| Dismissed | 0 | 0 | 0 | 0 |
| **Total** | **2** | **2** | **2** | **6** |

## Decisions

- Inline thread: `src/components/podcasts/episode-list.tsx:3` — Fixed: added `useMemo` to React imports.
- Inline thread: `src/components/podcasts/episode-list.tsx:56` — Fixed: wrapped filter in `useMemo`, moved hook above early returns.
- Review body: `pr_comment:4134487256` (gemini) — Answered: covered by the inline `useMemo` fixes.
- Review body: `pr_comment:4134512452` (coderabbit) — Deferred: renaming `Error`/`Loading`/`Empty` story exports would invalidate fresh VRT baselines; safe to keep as-is (no `throw new Error()` in the file). Can be done later as a project-wide story naming sweep.

## Follow-up

Author will address the CodeRabbit nitpick manually in a follow-up if the `*State` naming convention is adopted project-wide.

**Comment by funny-otter** [comment]
> 🧹 Nitpick comments (1)
> `src/components/podcasts/__tests__/episode-list.test.tsx` 151-162: Add an explicit `null` title regression case...

Fixed: added an explicit `title: null` fixture to the regression test and renamed it to cover both `undefined` and `null` cases, exercising `episode.title ?? ""`.

**Comment by funny-otter** [comment]
## PR Comment Status

| Status | Threads | Review Bodies | Issue Comments | Total |
|--------|---------|---------------|----------------|-------|
| Resolved | 2 | 2 | 2 | 6 |
| Fixed by DLC | 0 | 1 | 0 | 1 |
| Answered by DLC | 0 | 0 | 0 | 0 |
| Discussion-Deferred | 0 | 0 | 0 | 0 |
| **Total** | **2** | **3** | **2** | **7** |

## Changes in this cycle

- Review body `pr_comment:4134540045` (coderabbit) — **Fixed**: added explicit `title: null` fixture to the regression test so the `episode.title ?? ""` coercion is exercised for both `undefined` and `null`.
- **Out-of-band regression fix** (from `/codex:review --base main`): capped `BatchSummarizeButton` input to `BATCH_SUMMARIZE_LIMIT = 20` at the page-level call sites. Without this slice, the 200-episode load would hit `/api/episodes/batch-summarize`'s 20-episode cap and fail with a 400 on any podcast >20 episodes. Constant extracted to `src/lib/batch-summarize.ts` and imported by both the route and the page.

## Follow-up

None.
